### ISSUE TITLE
tf.while_loop swap_memory=True when not running train_op, does it always swap?

### ISSUE THREAD

**Issue by carlthome** [tensorflow/tensorflow#9720]
From the documentation of `tf.while_loop`:
> For training, TensorFlow remembers the tensors that are produced in the forward inference but needed in back propagation. These tensors can be a main source of memory consumption and often cause OOM problems when training on GPUs. When the flag swap_memory is true, we swap out these tensors from GPU to CPU. This for example allows us to train RNN models with very long sequences and large batches.

But **after** having trained, is TensorFlow's engine smart enough to avoid swapping even if `swap_memory=True` or should [optimize_for_inference.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference.py) flip swap_memory (i.e. make the equivalent modifications to the graph)?

Naturally this only applies to running inference on a GPU.

@yuanbyu

**Comment by concretevitamin** [comment]
@yuanbyu @hawkinsp could you take a look at the first question?  If the answer is no, we could reassign to someone to take a look at `optiomize_for_inference.py` or related tools.

**Comment by yuanbyu** [comment]
while_loop will not do any memory swapping if it is inference only. In training, it only performs swapping on tensors that are produced in the forward but used in the backprop.

**Comment by carlthome** [comment]
Thanks for clarifying! I hope this could be briefly mentioned in the documentation of tf.while_loop as well (e.g. that caching outputs during a forward pass is only done if tf.Session.run will be needing it for gradients and backprop) as it's not a clear behavior from a pure programming perspective in the more general case not related to machine learning.
